eclair-core/src/main/scala/fr/acinq/eclair/crypto/Sphinx.scala

```diff
@@ -74,298 +48,324 @@ object Sphinx extends Logging {
-    *                     was created or forwarded: see makePacket() and makeNextPacket()
-    * @param failure      failure message
-    * @return an error packet that can be sent to the destination node
-    */
-  def createErrorPacket(sharedSecret: ByteVector32, failure: FailureMessage): ByteVector = {
-    val message: ByteVector = FailureMessageCodecs.failureMessageCodec.encode(failure).require.toByteVector
-    require(message.length <= MaxErrorPayloadLength, s"error message length is ${message.length}, it must be less than $MaxErrorPayloadLength")
-    val um = Sphinx.generateKey("um", sharedSecret)
-    val padlen = MaxErrorPayloadLength - message.length
-    val payload = Protocol.writeUInt16(message.length.toInt, ByteOrder.BIG_ENDIAN) ++ message ++ Protocol.writeUInt16(padlen.toInt, ByteOrder.BIG_ENDIAN) ++ ByteVector.fill(padlen.toInt)(0)
-    logger.debug(s"um key: $um")
-    logger.debug(s"error payload: ${payload.toHex}")
-    logger.debug(s"raw error packet: ${(Sphinx.mac(um, payload) ++ payload).toHex}")
-    forwardErrorPacket(Sphinx.mac(um, payload) ++ payload, sharedSecret)
-  }
+    /**
+      * When an invalid onion is received, its hash should be included in the failure message.
+      */
+    def hash(onion: wire.OnionPacket): ByteVector32 =
+      Crypto.sha256(wire.LightningMessageCodecs.onionPacketCodec(onion.payload.length.toInt).encode(onion).require.toByteVector)
 
-  /**
-    *
-    * @param packet error packet
-    * @return the failure message that is embedded in the error packet
-    */
-  private def extractFailureMessage(packet: ByteVector): FailureMessage = {
-    require(packet.length == ErrorPacketLength, s"invalid error packet length ${packet.length}, must be $ErrorPacketLength")
-    val (mac, payload) = packet.splitAt(Sphinx.MacLength)
-    val len = Protocol.uint16(payload.toArray, ByteOrder.BIG_ENDIAN)
-    require((len >= 0) && (len <= MaxErrorPayloadLength), s"message length must be less than $MaxErrorPayloadLength")
-    FailureMessageCodecs.failureMessageCodec.decode(BitVector(payload.drop(2).take(len))).require.value
   }
 
   /**
-    *
-    * @param packet       error packet
-    * @param sharedSecret destination node's shared secret
-    * @return an obfuscated error packet that can be sent to the destination node
+    * A payment onion packet is used when offering an HTLC to a remote node.
     */
-  def forwardErrorPacket(packet: ByteVector, sharedSecret: ByteVector32): ByteVector = {
-    require(packet.length == ErrorPacketLength, s"invalid error packet length ${packet.length}, must be $ErrorPacketLength")
-    val key = generateKey("ammag", sharedSecret)
-    val stream = generateStream(key, ErrorPacketLength)
-    logger.debug(s"ammag key: $key")
-    logger.debug(s"error stream: $stream")
-    packet xor stream
-  }
+  object PaymentPacket extends OnionPacket {
+
+    override val PayloadLength = 1300
```

Comment:
It's not really halfway-through, because `PaymentPacket` isn't constrained to `PerHopPayload`s.
`PaymentPacket` is simply the packet that's used with `update_add_htlc` but it can either contain `PerHopPayload`s, variable-length payloads (tlv streams), or a mixture of both.

The way I view this Sphinx component is:

- it only deals with `bytes` (not typed payloads)
- `OnionPacket` implements the algorithm from the sphinx paper (where a packet starts with a version, then public key, then an onion-encrypted payload, then a mac)
- The sphinx paper doesn't constrain the size of the onion-encrypted payload at all, but it needs it to correctly generate the filler
- Payment packets will always have a fixed size of 1300 bytes
- Trampoline packets will always have a fixed size of 692 bytes
- If we implement Hornet at some point, we will want a lot of different packet sizes

I used that extension/inheritance scheme to be able to inject the `PayloadLength` value (necessary in all the computations inside `OnionPayload` methods) because the only other alternative I saw was to add the `payloadLength` as an argument of each function, which felt weird.

Once the computation is done it doesn't matter if the result is a trampoline onion or a payment onion as the higher layers should only treat it as a blob, so it makes sense that all returns an `OnionPacket` (imo).

Reply:
I do agree that the architecture may look complex, maybe it's because I'm not experience enough in Scala to find a better pattern...but we can refactor this later when we find one.